tket/src/WeightSubgrMono/Searching/NodeListTraversal.cpp

```diff
@@ -0,0 +1,289 @@
+// Copyright 2019-2022 Cambridge Quantum Computing
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "WeightSubgrMono/Searching/NodeListTraversal.hpp"
+
+#include <sstream>
+
+#include "Utils/Assert.hpp"
+#include "WeightSubgrMono/Common/GeneralUtils.hpp"
+#include "WeightSubgrMono/Searching/NodesRawData.hpp"
+
+namespace tket {
+namespace WeightedSubgraphMonomorphism {
+
+NodeListTraversal::NodeListTraversal(NodesRawDataWrapper& raw_data_wrapper)
+    : m_raw_data(raw_data_wrapper.m_raw_data) {}
+
+std::set<VertexWSM> NodeListTraversal::get_used_target_vertices() const {
+  std::set<VertexWSM> target_vertices;
+
+  // Examine all PV.
+  for (unsigned pv = 0; pv < m_raw_data.domains_data.size(); ++pv) {
+    const auto& domain_data = m_raw_data.domains_data[pv];
+
+    // Examine Dom(PV) at all levels.
+    for (unsigned level = 0; level <= domain_data.entries_back_index; ++level) {
+      const auto& node_index = domain_data.entries[level].node_level;
+      // Nogood nodes are ignored.
+      // So, check that at least one node with this domain
+      // is not a nogood; otherwise skip.
+      if (m_raw_data.nodes_data[node_index].nogood &&
+          level < domain_data.entries_back_index) {
+        const auto& next_node_index = domain_data.entries[level + 1].node_level;
+        bool some_node_is_good = false;
+        for (auto ii = node_index; ii < next_node_index; ++ii) {
+          if (!m_raw_data.nodes_data[ii].nogood) {
+            some_node_is_good = true;
+            break;
+          }
+        }
+        if (!some_node_is_good) {
+          // Skip.
+          continue;
+        }
+      }
+      const auto& domain = domain_data.entries[level].domain;
+      if (target_vertices.empty()) {
+        target_vertices = domain;
+      } else {
+        for (VertexWSM tv : domain) {
+          target_vertices.insert(tv);
+        }
+      }
+    }
+  }
+  return target_vertices;
+}
+
+bool NodeListTraversal::move_up() {
+  while (m_raw_data.current_node_level > 0) {
+    --m_raw_data.current_node_level;
+    if (m_raw_data.nodes_data[m_raw_data.current_node_level].nogood) {
+      continue;
+    }
+    for (unsigned pv = 0; pv < m_raw_data.domains_data.size(); ++pv) {
+      auto& domain_data = m_raw_data.domains_data[pv];
+      while (domain_data.entries[domain_data.entries_back_index].node_level >
```

Comment:
I will add a few comments and a bit more stuff in the README.

Unfortunately, this kind of complexity is unavoidable in C++, for performance reasons (I'm pretty sure - but I would be extremely interested if you find an error in my reasoning, or a simpler way of doing it which avoids the performance cost).

Maybe I should make a general template class which wraps some of this up; that would be easier to understand I think.

(There are similarities with src\TokenSwapping\include\TokenSwapping\VectorListHybrid.hpp...)

In src\WeightSubgrMono\include\WeightSubgrMono\Searching\NodesRawData.hpp we have struct DomainData which contains std::vector<Entry> entries.

LOGICALLY, "entries" simply contains the list of all data up to the present node index. (But note: we ONLY have an entry at a node where the domain CHANGES, hence there is NO relation between "node index" and the "entries_back_index"; they are indices for completely different vectors).

As we move up and down the search tree, "entries" would continually shrink and expand (data for larger node indices is outdated irrelevant junk which LOGICALLY should just be deleted).

Unfortunately, this would be very slow: each "struct Entry" object itself contains "std::set<VertexWSM> domain". Thus, constructing an Entry object involves heap memory allocation, not stack, which is expensive.

However, clearing an existing std::set and inserting elements is cheap, AS LONG AS the std::set was previously big enough that the previously allocated memory, which is still implicitly available and NOT deallocated, inside the std::set internal storage - which MIGHT internally be using a std::vector - can still contain all the new elements (in their std::set layout). 

Over time, as we keep clearing and reinserting, eventually this will be mostly true. (And actually will usually become true very quickly I think, as the domain sizes shrink as you move down the tree).

(OK, to be strict, I believe the C++ standard doesn't REQUIRE this of std::set; but it surely IS true in any good std::set implementation. It IS required by the standard that a std::vector reuses its previously allocated space efficiently when you clear and push back new elements).

(What I'm saying is, I BELIEVE that all sensible compilers internally will make std::set<T> use something similar to, or maybe even exactly, store most of their data as std::vector<A> internally, where "A" objects are T objects with extra added data, which keeps track of the tree structure necessary for a std::set to function).

Thus, we reuse "Entry" objects instead of deleting and then later reallocating them.entries_back_index is nothing more than what the index of the back() element of the std::vector<Entry> WOULD be, IF we actually resized and expanded this vector each time.

Also, note that entries_back_index and node_level are referring to indices in completely different vectors so they're not at all comparable.

(Note that clearing a std::vector<Entry> necessarily calls the destructor of Entry, which necessarily deallocates the memory, so it CANNOT be reused and reallocated quickly).

An alternative, which I think IS used a lot in games, is to write a CUSTOM new operator for Entry which actually has its own separate memory pool which DOESN'T get deallocated, and reuses the memory.

Anyway, maybe I'll make a template class which wraps this nicely, so we have functions called "resize", "clear" and "back" which work "logically rather than physically".

Finally, note that std::set<VertexWSM> is OK to reuse, BECAUSE VertexWSM is a primitive type - just an int type - and hence has trivial constructor/destructor. If VertexWSM were a complicated thing containing std::string, std::vector, etc. etc. then we'd need ANOTHER level of indirection and reuse!!!

Reply:
So basically, whenever you have a `std::vector<std::vector<int>>` which is resized a lot, inside a performance critical part, it's very likely that this is wasting a lot of time, and you need an alternative structure like `std::vector<ID>`, where `ID` is some simple struct (maybe just wrapping a `size_t`) which refers to ANOTHER container of `std::vector<int>` objects which doesn't actually erase any elements physically, just allows you to REUSE them.

Reply:
src\WeightSubgrMono\include\WeightSubgrMono\Common\SimpleStorage.hpp was doing one half of this: getting the new object; it is missing the other half: "logically" erase an element (which is not actually erased, but becomes available for reuse).

Anyway, this might well be coming soon in a new template class, which hopefully will be clearer!

Reply:
Actually, just done a very crude experiment, with clear and reuse vs. use new object:

With `std::vector<unsigned>` as the object, my approach above is MUCH faster: >10x speedup, confirming my expectation (the 10x figure a bit meaningless though, it of course depends on the size. I've used size 20 which is maybe typical of larger problems.

With `std::set<unsigned>`, my approach is ~10% faster. So still probably worth doing, but not as good as I expected. I suppose this just shows that the std::set insert operations are relatively slow.

Interesting, though, that with my approach, std::set is ~50 times slower than std::vector.

This is actually very encouraging because I'm planning to use dynamic bitsets instead of std::set, and of course dynamic bitsets are effectively just std::vector, so this should make all the domain parts MUCH faster than currently; how much will it speed it up overall? I'll guess 50%; we'll see!